nkms/crypto/powers.py

```diff
@@ -0,0 +1,122 @@
+from random import SystemRandom
+from typing import Iterable
+
+from py_ecc.secp256k1 import N, privtopub, ecdsa_raw_sign, ecdsa_raw_recover
+
+from nkms.crypto.crypto import Crypto
+from nkms.crypto.hash import signature_hash
+from nkms.crypto.keypairs import EncryptingKeypair
+
+
+class PowerUpError(TypeError):
+    pass
+
+
+class NoSigningPower(PowerUpError):
+    pass
+
+
+class NoEncryptingPower(PowerUpError):
+    pass
+
+
+class CryptoPower(object):
+    def __init__(self, power_ups=[]):
+        self._power_ups = {}
+        self.public_keys = {}  # TODO: The keys here will actually be IDs for looking up in a KeyRing.
+
+        if power_ups:
+            for power_up in power_ups:
+                self.consume_power_up(power_up)
+
+    def consume_power_up(self, power_up):
+        if isinstance(power_up, CryptoPowerUp):
+            power_up_class = power_up.__class__
+            power_up_instance = power_up
+        elif CryptoPowerUp in power_up.__bases__:
+            power_up_class = power_up
+            power_up_instance = power_up()
+        else:
+            raise TypeError("power_up must be a subclass of CryptoPowerUp or an instance of a subclass of CryptoPowerUp.")
+        self._power_ups[power_up_class] = power_up_instance
+
+        if power_up.confers_public_key:
+            self.public_keys[power_up_class] = power_up_instance.public_key()  # TODO: Make this an ID for later lookup on a KeyRing.
+
+
+    def pubkey_sig_bytes(self):
+        try:
+            return self._power_ups[SigningKeypair].pubkey_bytes()  # TODO: Turn this into an ID lookup on a KeyRing.
+        except KeyError:
+            raise NoSigningPower
+    def pubkey_sig_tuple(self):
+        try:
+            return self._power_ups[SigningKeypair].pub_key  # TODO: Turn this into an ID lookup on a KeyRing.
+        except KeyError:
+            raise NoSigningPower
+
+    def sign(self, *messages):
+        """
+        Signs a message and returns a signature with the keccak hash.
+
+        :param Iterable messages: Messages to sign in an iterable of bytes
+
+        :rtype: bytestring
+        :return: Signature of message
+        """
+        try:
+            sig_keypair = self._power_ups[SigningKeypair]
+        except KeyError:
+            raise NoSigningPower
+        msg_digest = b"".join(signature_hash(m) for m in messages)
```

Comment:
I am surprised that sign() can actually handle it (when you join multiple hashes). Or can it?
Anyway, I think it should probably be:
```
signature_hash(b"".join(signature_hash(m) for m in messages))
```